### DISCUSSION TITLE
Build Failure on WSL: Ubuntu 22.04 on x86_64 using ruby-build 20231107

### DISCUSSION THREAD

**Discussion by nzifnab** [rbenv/ruby-build#2298] [Build failures] [answered]
I have installed `rbenv` and `ruby-build` using homebrew and installed the dependencies using `apt-get` as described here: https://github.com/rbenv/ruby-build/wiki#ubuntudebianmint

## Environment Details

```
zifnab:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.2 LTS
Release:        22.04
Codename:       jammy
zifnab:~$ ruby-build --version
ruby-build 20231107
zifnab:~$ rbenv --version
rbenv 1.2.0
zifnab:~$ gcc --version
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

I can't seem to install any version of ruby using `rbenv` (Though this worked fine a few months ago).

When I try to install with `rbenv install 3.0.6 --verbose`, this is the full output:
https://gist.github.com/nzifnab/5515ecadf74aa2bae830cff962bd91c1

And here is the error in the log itself:

```
compiling symbol.c
compiling thread.c
during GIMPLE pass: pre
parse.y: In function ‘is_static_content’:
parse.y:11599:1: internal compiler error: Segmentation fault
11599 | is_static_content(NODE *node)
      | ^~~~~~~~~~~~~~~~~
0xd7a791 internal_error(char const*, ...)
        ???:0
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <file:///usr/share/doc/gcc-11/README.Bugs> for instructions.
compiling time.c
make: *** [Makefile:425: parse.o] Error 1
make: *** Waiting for unfinished jobs....
```

The error line itself changes each time I try to install, for instance, a 2nd run gives me the error here (rest is truncated for  brevity):

```
compiling time.c
compiling transcode.c
compiling transient_heap.c
during RTL pass: bbro
parse.y: In function ‘reduce_nodes’:
parse.y:11596:1: internal compiler error: in fixup_reorder_chain, at cfgrtl.c:3968
11596 | }
      | ^
0xd7a791 internal_error(char const*, ...)
        ???:0
0xd71eca fancy_abort(char const*, int, char const*)
        ???:0
Please submit a full bug report,
```

And yet another time, like this:

```
compiling thread.c
compiling time.c
compiling transcode.c
during RTL pass: *free_cfg
gc.c: In function ‘gc_update_object_references’:
gc.c:9297:1: internal compiler error: in insert_section_boundary_note, at bb-reorder.c:2619
 9297 | }
      | ^
0xd7a791 internal_error(char const*, ...)
        ???:0
0xd71eca fancy_abort(char const*, int, char const*)
        ???:0
Please submit a full bug report,
```

As you can see, each time seems to fail with a different error line, and I'm at a loss as to what's going on here.

When trying to install `rbenv install 3.2.2`, the installer will frequently hang instead of giving an error, here:

```
compiling enc/unicode.c
compiling enc/utf_8.c
compiling enc/trans/newline.c
warning: variable does not need to be mutable
   --> ./yjit/src/asm/mod.rs:627:17
    |
627 |             let mut cb = CodegenGlobals::get_inline_cb();
    |                 ----^^
    |                 |
    |                 help: remove this `mut`
    |
    = note: `#[warn(unused_mut)]` on by default

warning: variable does not need to be mutable
   --> ./yjit/src/asm/mod.rs:632:17
    |
632 |             let mut ocb = CodegenGlobals::get_outlined_cb().unwrap();
    |                 ----^^^
    |                 |
    |                 help: remove this `mut`

warning: variable does not need to be mutable
    --> ./yjit/src/core.rs:1487:13
     |
1487 |         let mut last_target = match &mut last_branch.targets {
     |             ----^^^^^^^^^^^
     |             |
     |             help: remove this `mut`

warning: 3 warnings emitted

touch yjit/target/release/libyjit.a
```

And then it just sits there not doing anything

When trying to use `rbenv install 3.2.2`

I've also tried locking the openssl version when installing ruby 3.2.2:

    RUBY_CONFIGURE_OPTS="--with-open-ssl-dir=$(brew --prefix openssl@3)" rbenv install 3.2.2 --verbose

But it's still hanging every time it hits that `touch yjit/target/release/libyjit.a` line.

**Comment by nzifnab** [comment, marked as answer]
I'm still unclear why the ruby `3.0.6` install was failing, but I have a working version of that in a different instance of WSL so I'm not going to worry about it... looking at that working instance, I suspect it has something to do with openssl:

```
zifnab:~$ ruby --version
ruby 3.0.6p216 (2023-03-30 revision 23a532679b) [x86_64-linux]
zifnab:~$ ruby -r openssl -e 'puts OpenSSL::OPENSSL_VERSION'
OpenSSL 1.1.1t  7 Feb 2023
```

1.1.1t... so when trying a modern install with openssl 3, is that why it's failing horribly and giving segfaults and the like?

I was able to get ruby 3.2.2 installed by using openssl 3.1 from homebrew and DISABLING yjit:

    RUBY_CONFIGURE_OPTS="--with-open-ssl-dir=$(brew --prefix openssl@3) --disable-yjit" rbenv install 3.2.2 --verbose

I'll be honest... I don't know what yjit does, and afaik we don't use rust, but I have no idea why it was hanging on that with no error. I left it for 2 hours and came back and it was still hung on that same line. This latest one _seems_ to have worked by disabling that

**Reply by mislav** [reply]
That's really weird. I cannot imagine what was the problem. Thanks for sharing your workaround!

**Reply by kalicki** [reply]
In my case this resolved:

`apt-get install -y libssl-dev && apt-get install -y libyaml-dev`

**Reply by osmanbowser** [reply]
Ran into the same issue on Ubuntu 22.04 but  while using ASDF.

Kept getting the error `BUILD FAILED (Ubuntu 22.04 on x86_64 using ruby-build 20231107)`

Turns out that I had curl installed incorrectly

Here's how I fixed it:

1. Uninstall snap curl  `sudo snap remove curl`
2. Reinstall curl properly `sudo apt install curl`
3. Update apt-get `sudo apt-get update`
4. Install the necessary packages (referencing [GoRails](https://gorails.com/setup/ubuntu/22.04)] `sudo apt-get install git-core zlib1g-dev build-essential libssl-dev libreadline-dev libyaml-dev libsqlite3-dev sqlite3 libxml2-dev libxslt1-dev libcurl4-openssl-dev software-properties-common libffi-dev`
5. Add the necessary plugins for Rails `asdf plugin add ruby` and `asdf plugin add nodejs` (most likely unrelated to the issue but needed for Rails)
6. Install ruby using `asdf install ruby 3.2.2`

The build succeeded and I was able to finish installing Rails.

Hope this helps!

**Reply by KelsonF** [reply]
It worked, thanks for the help =)

**Reply by nzifnab** [reply]
Another solution that worked for me to resolve the SEGFAULT was to change the number of `MAKE_OPTS`. `ruby-build` was defaulting to `make -j 32`, but if I use `MAKE_OPTS=-j2 rbenv install 3.3.1` then it suddenly works. I wonder if the number of CPU cores in my WSL environment is not as many as it thinks.

**Reply by mateuseap** [reply]
Thanks @osmanbowser, worked for me in Ubuntu 24.04!

Here's the build error for record:
```bash
==> Installing openssl-3.0.14...
-> ./config "--prefix=$HOME/.asdf/installs/ruby/3.3.2/openssl" "--openssldir=$HOME/.asdf/installs/ruby/3.3.2/openssl/ssl" zlib-dynamic no-ssl3 shared

BUILD FAILED (Ubuntu 24.04 on x86_64 using ruby-build 20240612)
```

**Reply by zigpot** [reply]
> apt-get install -y libssl-dev && apt-get install -y libyaml-dev

this one worked for me

_-Ubuntu 22.04.4 LTS (Jammy Jellyfish)_

**Reply by Justpunk86** [reply]
Man @osmanbowser! Thanks for your help!
This worked for me on Ubuntu 22.04:
> 4\. sudo apt-get install git-core zlib1g-dev build-essential libssl-dev libreadline-dev libyaml-dev libsqlite3-dev sqlite3 libxml2-dev libxslt1-dev libcurl4-openssl-dev software-properties-common libffi-dev

**Reply by sandbardev** [reply]
Looking at the logs indicated at the error message, I noticed I was missing a C compiler on my WSL environment.
I managed to get it to work by installing the `libyaml-dev libssl-dev build-essential` dependencies.

**Reply by mikodagatan** [reply]
Doing this worked for me. 

`sudo apt-get update`
`sudo apt install build-essential rustc libssl-dev libyaml-dev zlib1g-dev libgmp-dev`

**Reply by surajw-swivel** [reply]
worked 
sudo apt-get update
sudo apt install build-essential rustc libssl-dev libyaml-dev zlib1g-dev libgmp-dev

**Reply by Yasushiki** [reply]
> Doing this worked for me.
> 
> `sudo apt-get update` `sudo apt install build-essential rustc libssl-dev libyaml-dev zlib1g-dev libgmp-dev`

Worked
